### PR TITLE
Datasources as configuration

### PR DESCRIPTION
**Pull request by bergquist**
Adds support for configuring datasources from files. When Grafana starts it will read the config file and insert/update/delete datasources based on the configuration file. 

I would like to add support for flagging datasources in the UI as read only. So that they are only configurable from config files.

Example
```yaml
# list of datasources that should be deleted from the database
delete_datasources:
   - name: Graphite
     org_id: 1
# list of datasources to insert/update depending 
# whats available in the datbase
datasources:
    # <string, required> name of the datasource. Required
  - name: Graphite
    # <string, required> datasource type. Required
    type: graphite
    # <string, required> access mode. direct or proxy. Required
    access: proxy
    # <int> org id. will default to org_id 1 if not specified
    org_id: 1
    # <string, required> url
    url: http://localhost:8080
    # <string> database password, if used
    password:
    # <string> database user, if used
    user:
    # <string> database name, if used
    database:
    # <bool> enable/disable basic auth
    basic_auth:
    # <string> basic auth username
    basic_auth_user:
    # <string> basic auth password
    basic_auth_password:
    # <bool> enable/disable with credentials headers
    with_credentials:
    # <bool> mark as default datasource. Max one per org
    is_default:
    # <string> json data
    json_data: '{"graphiteVersion":"0.9"}'
    # <string> json object of data that will be encrypted in UI.
    secure_json_fields: ''
    # <int> including this value garantees that instance with old configs cannot
    #       overwrite your last change.
    version: 1

```

**TODO**
- [x] Basic implementation of datasources from cfg
- [x] support deleting datasources not in config file
- [x] Make sure this works when 2+ grafana servers use the same DB. 
  - [x] 2+ servers starting at the same time.
  - [x] 2+ servers with different configs starting at the same time.
  - [x] support optimistic concurrency for data source updates.
- [x] Docs
  - [x] improve `conf/datasources.yaml` example
- [x] Datasources from cfg file should be read only in UI
- [x] rename read-only to editable.
- [x] Support loading multiple configfiles. `/conf/datasources/*.yaml`
- [x] JsonData should be yaml instead of json.
- [x] More docs for each datsource what is expected to be stored in JsonData

image

image

**Running multiple grafana instances**
* Two Grafana servers sharing the same config (network share) will be fine since both acts on the same config.
* Two servers with different configs due to deploy/split brain can end up overwriting each others datasources. To avoid this we include the version setting for each datasources. By incrementing this value by hand for each change it will be garanteed to overwrite old values.

ref
https://github.com/grafana/grafana-docker/issues/11
https://github.com/grafana/grafana-docker/pull/98
https://github.com/grafana/grafana/issues/1789

### CODE REVIEW

**Review by torkelo — APPROVED** [review]

File: pkg/setting/datasources/types.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		Password:          ds.Password,
+		User:              ds.User,
+		Database:          ds.Database,
+		BasicAuth:         ds.BasicAuth,
+		BasicAuthUser:     ds.BasicAuthUser,
+		BasicAuthPassword: ds.BasicAuthPassword,
+		WithCredentials:   ds.WithCredentials,
+		IsDefault:         ds.IsDefault,
+		JsonData:          jsonData,
+		SecureJsonData:    ds.SecureJsonData,
```
Comment (torkelo) [inline pkg/setting/datasources/types.go]: where do you encrypt the secureJsonData?

Comment (torkelo) [reply pkg/setting/datasources/types.go]: no matter, it's done in the sqlstore

File: pkg/cmd/grafana-server/server.go [hunk 2]
```diff
@@ -55,6 +57,12 @@ func (g *GrafanaServerImpl) Start() {
 	g.writePIDFile()
 
 	initSql()
+	err := dsSettings.Init(filepath.Join(setting.HomePath, "conf/datasources.yaml"))
```
Comment (torkelo) [inline pkg/cmd/grafana-server/server.go]: think I would prefer this as a service and initialized like: 

```go
setupFromFilesSrv := setupFromFilesSrv.New(setting.HomePath, "conf/datasources.yaml")
setupFromFilesSrv.childRoutines.Go(func() error { return engine.Run(g.context) })
```

Thinking is that we might want Grafana to detect changes to these files and sync without restarting Grafana. Aslo naming the service something like setupFromFilesSrv (or a better name) makes it more open to be used for Users etc.

Comment (bergquist) [reply pkg/cmd/grafana-server/server.go]: I don't want to include auto reloading in the scope of this PR. But once this PR is merged it should be easy to start working on that.

Comment (torkelo) [reply pkg/cmd/grafana-server/server.go]: I am fine with that, but still think it should be a service  you create, and not a static package Init function as we are trying to move away from that

Comment (bergquist) [reply pkg/cmd/grafana-server/server.go]: You right.

The example code you posted starts the process as a go routine. Wouldn't you want this to be blocking? If the setup fails, shouldn't grafana start up fail? I can see arguments for both answers. But I lean towards start up failing cause if you configure data source this way you really expected those to exist.

Comment (torkelo) [reply pkg/cmd/grafana-server/server.go]: Yes, your right about that. Should block startup

**Comment by bergquist** [comment]
Some feedback from https://twitter.com/fredbrancz
> What we currently do in the sidecar is that we actually glob datasource files, so it would be great if that lands in upstream as well :)

https://twitter.com/fredbrancz/status/923162462572175360

**Comment by TheBigfoot** [comment]
Pretty awesome feature, but one question:
Will there be a parameter for assigning it to a specific organization? Right now it looks like it will only assign it to the default organization? I would be also fine with assigning it to all, but only the first org may be a problem.

**Comment by bergquist** [comment]
@TheBigfoot seems like I forgot to add org_id parameter to the example. If you specify a none zero value Grafana will insert the datasource in that org. 

If the value is 0 or missing Grafana will insert it the first org.

**Comment by bergquist** [comment]
With support for globbing, it doesn't make sense to have `purge_other_datasources` on DatasourcesAsCfg anymore. But we still need to be able to delete datasources from config files.

Currently, I see two options.
* Move purge_other_datasources to grafana.ini
* Remove it and add support for deleting dashsources another way. One suggestion would be to introduce a new array in the config file with dashboards to delete. ex
```yaml
delete_datasources:
  - name: prometheus
    org_id: 3
datasources:
  - name: Graphite
    type: graphite
    access: proxy
    org_id: 3

```
Any feedback on either idea?

**Comment by botzill** [comment]
Hi, I see that it's a work in progress, any estimations when we will be able to have a nightly build to test this feature?

**Comment by bergquist** [comment]
@botzill 1-2 weeks. There are still some things I want to improve before merging this to master.

**Comment by botzill** [comment]
Thx @bergquist! 

Waiting to test this out.

**Comment by jeremyeder** [comment]
@mrsiano

**Comment by mrsiano** [comment]
@jeremyeder 
@bergquist, is it give us the ability to config my datasource as i wish, or this config file followed by template ? 
also, when loading that config file grafana will check for args validation ?

**Comment by bergquist** [comment]
@mrsiano This PR will make it possible to configure data sources from a file instead of the UI. The main purpose of this is to make provisioning easier.

**Comment by mrsiano** [comment]
@bergquist sure thing, I just wonder if the config file could get any args that you wish.
I just reviewed the https://github.com/grafana/grafana/pull/9504/commits/87983021e240b3cad079c00d7ef6515872145fcb

so looks like yes.
this patch could help us to tweak the jsondata to use some token.

**Comment by bergquist** [comment]
@daniellee could you review the improved docs? I think this is understandable enough? I didn't want to scatter the settings to other datasource pages since its only gonna be used by those who provision datasources using the config file. 

After that, this is ready to be merged.

**Comment by bsideup** [comment]
🎉

**Comment by mrsiano** [comment]
@bergquist since this PR merged, can I use docker pull grafana:latest and use it ?

**Comment by bergquist** [comment]
@mrsiano no. `latest` is the last stable release. but you can use `master` to get this feature.

**Comment by tttMelody** [comment]
@bergquist awesome! When will the next release release? Thanks.

**Comment by bergquist** [comment]
@tttMelody we currently don't have a date for the next release. But you can install nightly now if you want to use data sources as configuration. The nightly build is fairly stable.

**Comment by tttMelody** [comment]
@bergquist Got it! Thanks.

**Comment by tttMelody** [comment]
Hi,
Is this can be supported?
image

image

**Comment by bergquist** [comment]
@tttMelody I dont see us supporting templating for those config files. But there might be other configuration management tools that can solve that for you. chef/ansible/puppet/saltstack.

Perhaps you could use localhost if you pointing at the same hostname where grafana is running?

**Comment by tttMelody** [comment]
Hi @beejeebus Thanks a lot to your advise.

In [configuration](http://docs.grafana.org/installation/configuration/), seems can do that,
image

I deploy grafana and influxDB in one node, and if I use localhost, I found that grafana will try to connect the influxDB in my computer(where my browser in), not the computer that I deploy the influxDB.

**Comment by botzill** [comment]
Hi,

I'm testing grafana version `Version 4.7.0-pre1 (commit: 03c23dd)` and when I set the `json_data` field it does not start, example:

```
- name: "PostgreSQL"
    type: postgres
    org_id: 1
    access: "proxy"
    url: "172.16.1.30:5432"
    password: "test"
    user: "test"
    database: "test"
    is_default: false
    json_data: '{"sslmode":"disable"}'
    secure_json_fields: ''
```

When I comment `json_data` it starts OK, when I leave it I get:

```
t=2017-12-14T12:51:15+0100 lvl=info msg="Shutdown started" logger=server code=1 reason="Startup failed"
```
I do smth wrong?

**Comment by bergquist** [comment]
@botzill we updated the json_data field to be yaml as well to make it more consistent. ex
http://docs.grafana.org/administration/provisioning/#example-datasource-config-file

**Comment by botzill** [comment]
Thx @bergquist, really helpful. Once more question, how to I set password from config file? I see it's not set when opening the datasource in UI. I have a simple pass like: `password: "test"`.

Thx!

**Comment by botzill** [comment]
OK, I see this as well, sorry :). We should use `secure_json_data` and set the `user` `password` for postgresql there. 

Now all works, thx!

**Comment by desertkun** [comment]
@bergquist Sorry for bothering, but do you have the information on when this will be released?

**Comment by botzill** [comment]
Hi,

Seems that we don't have here http://docs.grafana.org/administration/provisioning/#example-datasource-config-file specified the `tlsSkipVerify` option or maybe it was added somewhere ?

**Comment by bergquist** [comment]
@botzill fixed! will deploy the docs later. `tlsSkipVerify` should be added to `jsonData`

@desertkun this will be released in 5.0. We will release a beta within a few weeks.

**Comment by drauschenbach** [comment]
I'm struggling to figure out how to use a datasource file using the Docker container, which only exposes `/etc/grafana/` and `/var/lib/grafana/` volumes, but not the <app>/conf/ directory. This PR is said to close https://github.com/grafana/grafana-docker/issues/11, but after many hours I can't find a way to get it work, or any docs that would help.

**Comment by b2ronn** [comment]
Can I add a one datasource from the config file to all the organizations?

**Comment by bergquist** [comment]
@b2ronn No. You have to add the datasource once per org.

**Comment by mrsiano** [comment]
@bergquist is it possible to inject configuration or datasource to grafana container ?!
see also this:
https://github.com/grafana/grafana-docker/issues/124

**Pull request merged**